docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,488 @@
+
+### Tenets
+
+#### Same UX
+
+Cluster API user experience MUST be the same when using default, immutable updates or when using rollout extensions: e.g. in order to trigger a MachineDeployment rollout, you have to rotate a template, etc.
+
+#### Fallback to Immutable rollouts
+
+If rollout extensions can not cover the totality of the desired changes, users SHOULD be able to defer to Cluster API’s default, immutable rollouts. This is important for a couple of reasons:
+
+* It allows to implement custom rollout strategies incrementally, without the need to cover all use cases up-front.
+* There are case when replacing the machine will always be necessary:
+    * When it is not possible to recover the machine, e.g. hardware failure.
+    * When the user determines that recovering the machine is too complex/costly vs replacing it. 
+    * Automatic machine remediation (unless you use external remediation strategies)
+
+#### Clean separation of concern
+
+The rollout extension will be responsible to perform the updates on a single machine.
+
+The responsibility to determine which machine should be rolled out as well as the responsibility to handle rollout options like MaxSurge/MaxUnavailable will remain on the controllers owning the machine (e.g. KCP, MD controller).
+
+### Goals
+
+- Enable the implementation of in-place update strategies.
+- Allow users to update Kubernetes clusters using pluggable External Update Extension.
+- Maintain a coherent user experience for both rolling and in-place updates.
+- Support External Update Extensions for both Control Plane (KCP or others) and MachineDeployment controlled machines.
+- Allow in-place updates for single-node clusters without the requirement to reprovision hosts.
+
+### Non-Goals
+
+- To provide rollbacks in case of an in-place update failure. Failed updates need to be fixed manually by the user on the machine or by replacing the machine.
+
+## Proposal
+
+We propose a pluggable update strategy architecture that allows External Update Extension to handle the update process. The design decouples core CAPI controllers from the specific extension implementation responsible for updating a machine. The External Update Strategy will be configured reusing the existing field in KCP and MD resources, by introducing new type of strategy called `ExternalUpdate` (reusing the existing field in KCP and MD). This allows us to provide a consistent user experience: the interaction witht he CAPI resources is the same as in rolling updates.
+
+This proposal introduces a Lifecycle Hook named `ExternalUpdate` for communication between CAPI and external update implementers. Multiple external updaters can be registered, each of them only covering a subset of machine changes. The CAPI controllers will ask the external updaters what kind of changes they can handle and, based on the reponse, compose and orchestrate them to achieve the desired state.
```

Comment:
The proposal is a missing details in how the external updater logic would work, and how the "kind of changes they can handle" is handled. How is that going to work?

Reply:
I think It'd be good for the proposal to include a reference external updater implementation and shape around one common/trivial driving use case. E.g perform an in-place rolling update of the kubernetes version for a pool of Nodes. Then we can grasp and discuss design implications for RBAC, drain...

Reply:
@enxebre In the 'test plan' section we mention a "CAPD Kubeadm Updater", which will be a reference implementation and also used for testing.

Reply:
@vincepri

What do you mean with "how is that going to work?"? Are you referring to how the external updater knows what are the desired changes? Or how does the external updater compute what changes it can perform and what changes it can't?

Trying to give a generic answer here, the external updater will receive something like "current state" and "desired state" for a particular machine (including machine, infra machine and bootstrap) in the `CanUpdateRequest`. Then it will respond with something like an array of fields for those objects (kubeadmconfig -> `["spec.files", "spec.mounts", "spec.files"]`), which would signal the subset of fields that it can update.

Reply:
@enxebre 
The idea of opening the draft at this stage for review is to get feedback on the core ideas and high level flow before we invest more time on this direction. Unless you think that a reference implementation is necessary to have these discussions, I would prefer to avoid that work.

That said, I totally get that it's possible that the lack of detail in certain areas is making difficult to have the high level discussion. If that's the case, we are happy to add that detail wherever needed.

Reply:
> Trying to give a generic answer here, the external updater will receive something like "current state" and "desired state" for a particular machine (including machine, infra machine and bootstrap) in the CanUpdateRequest. Then it will respond with something like an array of fields for those objects (kubeadmconfig -> ["spec.files", "spec.mounts", "spec.files"]), which would signal the subset of fields that it can update.

These details must be part of the proposal, the details on how the entire flow from MachineDeployment, to the external request, back to the Machine, and reflecting status are not present, which makes it hard to understand how the technical flow will go and/or propose alternative solutions.